### PR TITLE
[QeRL] Add warnings for extra memory buffering 

### PR DESCRIPTION
**Pull request by kylesayrs**
## Background ##
Layerwise reloading allows users to incrementally load and process weights as they are loaded into the model. This system relies on buffering layer weights on device until all weights of a layer have been loaded. However, without CPU offloading, this approach necessarily causes excessive buffering if weights are loaded out of order.

Ideally, weights are loaded "in order", meaning that each layer's weights are fully loaded before loading the next layer's weights.
Screenshot 2026-04-19 at 20 48 38

However, loading "out of order" can cause layer weights to stay buffered while other layer weights are loading, causing excessive memory usage. In this example, q_proj, k_proj, v_proj, and up_proj are all buffered at the same time.
Screenshot 2026-04-19 at 20 48 47

## Purpose ##
* Log warnings if a user loads weights out of order. These warnings suggest that the user load weights in order

## Changes ##
* Add weakset `LOADING_LAYERS` which tracks which layers are being loaded at a time. If more than two layers are being loaded at the same time, a warning triggers
  * This warning will trigger when weights are loaded out of order
  * This warning will also trigger if weights have padding, which is one of the known flaws of the layerwise system

## Testing ##
* Tested using `test_online_quantize_reload`, confirmed that warnings do not trigger when loading from disk/cpu but do trigger when loading from gpu via patch.

### CODE REVIEW

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 1]
```diff
@@ -174,11 +182,27 @@ def online_process_loader(*args, **kwargs):
...
+                    str(list(names)),
+                )
+
         # Process and copy when all weights are loaded
-        if info.load_numel >= info.load_numel_total and not isinstance(  # type: ignore[operator]
-            layer, (Attention, MLAAttention)
-        ):
+        if info.load_numel >= info.load_numel_total:  # type: ignore[operator]
             _layerwise_process(layer, info)
+            if layer in LOADING_LAYERS:
```
Comment (Josephasafg) [inline vllm/model_executor/model_loader/reload/layerwise.py:204]: nit: `LOADING_LAYERS.discard(layer)`?

File: vllm/model_executor/model_loader/reload/utils.py [hunk 2]
```diff
@@ -39,3 +41,11 @@ def get_layer_size(layer: torch.nn.Module) -> int:
         for name, tensor in get_layer_tensors(layer).items()
         if name not in SKIP_TENSORS
     )
+
+
+def has_device_tensors(bound_args: BoundArguments) -> bool:
+    return any(
+        isinstance(value, torch.Tensor)
```
Comment (Josephasafg) [inline vllm/model_executor/model_loader/reload/utils.py:48]: nit: maybe a single check -
```python
  return any(
      isinstance(value, torch.Tensor) and value.device.type not in ("meta", "cpu")
      for value in bound_args.arguments.values()
  )
```

File: vllm/model_executor/model_loader/reload/layerwise.py [hunk 3]
```diff
@@ -174,11 +182,27 @@ def online_process_loader(*args, **kwargs):
...
+        # Do not online process attention layers, must wait until finalize
+        if isinstance(layer, (Attention, MLAAttention)):
+            return ret
+
+        # Log warnings allocating excessive buffers on device
+        if has_device_tensors(bound_args):
+            LOADING_LAYERS.add(layer)
+            if len(LOADING_LAYERS) >= 2:
+                names = sorted([layer.__class__.__name__ for layer in LOADING_LAYERS])
+                logger.warning_once(
```
Comment (Josephasafg) [inline vllm/model_executor/model_loader/reload/layerwise.py:194]: Should we also log the memory overhead in MB/GB? I think it could be informative. what do you think?

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/layerwise.py:198]: I agree, I've added this calculation

File: vllm/model_executor/model_loader/reload/utils.py [hunk 4]
```diff
@@ -44,8 +46,28 @@ def get_layer_size(layer: torch.nn.Module) -> int:
...
+    """
+    Calculate the number of bytes used by loaded weights for a given layer
+
+    :param info: layerwise info to get size of
+    :return: number of bytes used by loaded weights
+    """
+    return sum(
+        value.nbytes
+        for _, args in info.loaded_weights
+        for value in args.arguments.values()
```
Comment (Josephasafg) [inline vllm/model_executor/model_loader/reload/utils.py:71]: can we use `has_device_tensors`?

Comment (kylesayrs) [reply vllm/model_executor/model_loader/reload/utils.py:71]: Technically no, since some of the arguments might not be tensors/ might not be device tensors. I tried this and I found it's better to write explicitly here

**Review by Josephasafg — APPROVED** [review]

**Review by sfeng33 — APPROVED** [review]

**Comment by kylesayrs** [comment]
These tests are also failing on main as well
https://buildkite.com/vllm/ci/builds/63249
